crates/components/cipher/src/lib.rs

```diff
@@ -0,0 +1,381 @@
+    ///
+    /// # Arguments
+    ///
+    /// * `vm` - The necessary virtual machine.
+    /// * `explicit_nonce` - The TLS explicit nonce.
+    /// * `start_ctr` - The TLS counter number to start with.
+    /// * `message` - The message to en-/decrypt.
+    pub fn assign(
+        self,
+        vm: &mut dyn Vm<Binary>,
+        explicit_nonce: <<C as CipherCircuit>::Nonce as Repr<Binary>>::Clear,
+        start_ctr: u32,
+        message: Input,
+    ) -> Result<Vector<U8>, CipherError>
+    where
+        <<C as CipherCircuit>::Counter as Repr<Binary>>::Clear: From<[u8; 4]>,
+        <<C as CipherCircuit>::Nonce as Repr<Binary>>::Clear: Copy,
+    {
+        let (len, message) = match message {
+            Input::Message(msg) => (msg.len(), Some(msg)),
+            Input::Length(len) => (len, None),
+        };
+
+        self.assign_nonce_and_ctr(vm, explicit_nonce, start_ctr, len)?;
+
+        if let Some(msg) = message {
+            vm.assign(self.input, msg).map_err(CipherError::new)?;
+        }
+        vm.commit(self.input).map_err(CipherError::new)?;
+
+        Ok(self.output)
+    }
+
+    /// Assigns nonces and counters but reuses already assigned references for
+    /// the message inputs.
+    ///
+    /// # Arguments
+    ///
+    /// * `vm` - The necessary virtual machine.
+    /// * `explicit_nonce` - The TLS explicit nonce.
+    /// * `start_ctr` - The TLS counter number to start with.
+    /// * `len` - The length of the message to en-/decrypt.
+    pub fn assign_reuse(
+        self,
+        vm: &mut dyn Vm<Binary>,
+        explicit_nonce: <<C as CipherCircuit>::Nonce as Repr<Binary>>::Clear,
+        start_ctr: u32,
+        len: usize,
+    ) -> Result<Vector<U8>, CipherError>
+    where
+        <<C as CipherCircuit>::Counter as Repr<Binary>>::Clear: From<[u8; 4]>,
+        <<C as CipherCircuit>::Nonce as Repr<Binary>>::Clear: Copy,
+    {
+        self.assign_nonce_and_ctr(vm, explicit_nonce, start_ctr, len)?;
+        Ok(self.output)
+    }
+
+    /// Returns the stream length in bytes.
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&self) -> usize {
+        self.input.len()
+    }
+
+    fn assign_nonce_and_ctr(
+        &self,
+        vm: &mut dyn Vm<Binary>,
+        explicit_nonce: <<C as CipherCircuit>::Nonce as Repr<Binary>>::Clear,
+        start_ctr: u32,
+        len: usize,
+    ) -> Result<(), CipherError>
+    where
+        <<C as CipherCircuit>::Counter as Repr<Binary>>::Clear: From<[u8; 4]>,
+        <<C as CipherCircuit>::Nonce as Repr<Binary>>::Clear: Copy,
+    {
+        if self.len() != len {
+            return Err(CipherError::new(format!(
+                "message has wrong length, got {}, but expected {}",
+                len,
+                self.len()
+            )));
+        }
+
+        let message_len = len as u32;
+        let block_count = (message_len / 16) + (message_len % 16 != 0) as u32;
```

Comment:
Did you mean to use
`let block_size = <<C as CipherCircuit>::Block as StaticSize<Binary>>::SIZE / 8;`
instead of hard-coding 16 ?

Reply:
good catch :+1:  yeah that would have been the cleaner solution. But I think that code has been superseded by now.